accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -591,6 +591,15 @@ It is a compile-time error if:
 
 *   Any of the entry key expressions are not constant expressions.
 
+*   If any two keys in the map are identical. *Map patterns that don't have a
```

Comment:
I just found out from @johnniwinther that this is difficult to do in the CFE, because the CFE does constant evaluation as a separate step, after lowering everything to a kernel representation, and the kernel representation doesn't preserve patterns (because they get lowered to imperative code).

Could we make this an analyzer hint rather than a warning (kind of like we do for non-const maps)?  That way the user still finds out about the problem, but we don't have to pay the CFE implementation cost.

Reply:
The spec already states that it's an error if two map keys both have primitive equality and their `==` returns `true`, which also involves constant evaluation. Does that one worry them too?

The language also already has a similar check for constant map literals. Is there a reason this one is harder?

I'm not attached to it either way, I'm just a little surprised it's a problem given that the very next bullet point also requires it and so far no one has said it was a challenge.

Reply:
If the pattern lowering (desugaring) removes the information that these constants were map pattern keys of the same pattern, then it's harder than for a map literal, which is not desugared.

So the main problem is that desugaring happens before constant evaluation, but comparing keys for same-ness happens after constant evaluation, at which point it's no longer clearly visible which things must be non-seme.

(I'm a little impressed that we can lose information like that, and still be able to determine exhaustiveness. I guess the only constants needed for exhaustiveness are enum constants, which can be handled symbolically.)

Reply:
> (I'm a little impressed that we can lose information like that, and still be able to determine exhaustiveness. I guess the only constants needed for exhaustiveness are enum constants, which can be handled symbolically.)

Yeah, the existing (pre-patterns) exhaustiveness logic only recognizes direct references to enum values.  If your switch case refers to a constant, and that constant evaluates to the enum value, exhaustiveness checking doesn't handle it properly.  So maybe it's not so impressive as you thought 🥲.

Reply:
This is a bit concerning for general pattern exhaustiveness checking.  It seems to me that we'll need constants as part of exhaustiveness checking, and I don't imagine we'll do exhaustiveness post-lowering.

Reply:
I've added an item to the agenda to discuss this at the next shared patterns infrastructure meeting.

Reply:
The problem is not only the lowering but the fact that identity of constants is known very late in the analysis/compilation process. To support the keys check, and more importantly the exhaustiveness check on constants, we need to ensure that such a time even exists. For this I propose the following:

1. The value of constants (including identities) must never effect type inference/checking.
2. Unknown constants (`.fromEnvironment` constants) must be defined to have an unknown constant value for type inference/checking, "redundancy checking" (like this check for disjoint keys) and exhaustiveness checking.

1) is needed to allow the handling of constant evaluation to be performed fully after inference and I believe that it is already the case; for instance the type of a switch expression does not take the reachability of cases into account. 2) is needed to ensure that redundancy and exhaustiveness is well-defined for unknown constants for both the analyzer (which never knows their value) and the CFE (which only sometimes knows their value).

Reply:
If we _require_ exhaustiveness of enum value switches, then we can assume it during inference, and check it later. 
Might only works for straight switches on the enum type, not, e.g., on `(EnumType, int)`. (Do we require them to be exhausted if the rest of the pattern isn't exhaustive?)

The only other constants that we might do exhaustiveness checking on are booleans and `null`. I don't think we have other guaranteed finite types.

It's still worrisome that we can't prove exhaustiveness of:
```dart
const hasFoo = bool.hasEnvironment("foo");
const noFoo = !hasFoo;
...
switch (boolXFoo) {
  case (hasFoo, var x): ...
  case (noFoo, var x): ...
}
```
I can probably live with "environment"-values being considered too dynamic to be known in switches, but it is a special-case, and it can propagate through multiple constant expressions.

If we fail to show exhustiveness, end it leads to an error, and a "too dynamic" constant is involved in that, we should see if we can mention that in the error message.

Reply:
> The value of constants (including identities) must never effect type inference/checking.

Yeah, I don't think that's a problem. As Lasse says, inference/checking can basically assume that a switch *is* exhaustive if it must be and then we just report a compile-time exhaustiveness error later if that turns out to not be the case.

> Unknown constants (.fromEnvironment constants) must be defined to have an unknown constant value for type inference/checking, "redundancy checking" (like this check for disjoint keys) and exhaustiveness checking.

That's definitely an odd corner of the language, but it should be doable. We already need to deal with guards, which is another place where a pattern may not fully participate in exhaustiveness checking because we can't be sure at compile time that it will match.

> (Do we require them to be exhausted if the rest of the pattern isn't exhaustive?)

No:

> We define an *exhaustive type* to be:
>
> ...
>
> *   A record type whose fields are all exhaustive types

So a record type that has any non-exhaustive fields is not required to be exhaustively covered in a switch statement or switch element. Switch expressions must always be exhaustive.

> It's still worrisome that we can't prove exhaustiveness of:
>
> ```dart
> const hasFoo = bool.hasEnvironment("foo");
> const noFoo = !hasFoo;
> ...
> switch (boolXFoo) {
>   case (hasFoo, var x): ...
>   case (noFoo, var x): ...
> }

It's weird, but people don't use environment flags often anyway. You can think of the above example as more or less the same as:

```dart
const hasFoo = bool.hasEnvironment("foo");
const noFoo = !hasFoo;
...
switch (boolXFoo) {
  case (var y, var x) when y == hasFoo: ...
  case (var y, var x) when y == noFoo: ...
}
```

The environment flags are sort of not really constants.

So... how should I resolve this PR? I'm not attached to any particular choice here but I do think we should at least be internally consistent. If primitive `==` keys in a map pattern is a compile-time error then I don't see why identical keys shouldn't be too. If it's too hard to make identical keys and error, then we probably shouldn't make primitive `==` keys an error either.

Reply:
OK, I see that the PR has LGTMs, so I'm going to go ahead and merge this in. I'm happy to iterate on it more if needed afterwards. :)